snuba/utils/streams/configuration_builder.py

```diff
@@ -10,19 +10,28 @@
 from snuba.utils.streams.types import KafkaBrokerConfig
 
 
-def _get_default_topic_configuration(topic: Optional[Topic]) -> Mapping[str, Any]:
+def _get_default_topic_configuration(
+    topic: Optional[Topic], slice_id: Optional[int] = None
+) -> Mapping[str, Any]:
+
     if topic is not None:
-        return settings.KAFKA_BROKER_CONFIG.get(topic.value, settings.BROKER_CONFIG)
+        if slice_id is not None:
+            return settings.SLICED_KAFKA_BROKER_CONFIG.get(
+                (topic.value, slice_id), settings.BROKER_CONFIG
```

Comment:
Please make slice=0 behave the same as slice=None. The subtle differences here make it too easy to pass the wrong values.

Reply:
This does indeed work the same way between slice=0, slice=1, etc. for sliced topics. If you look into `settings/__init__.py`, you will see under the Slicing Configuration section, all of these configs/maps are for sliced topics/storages, i.e. those that are associated with multiple slices. So, in this code snippet, as long as the topic being passed in is one that is indeed defined to be "sliced", `SLICED_KAFKA_BROKER_CONFIG` will contain information for it's slice 0, slice 1, and so on. I hope that makes sense

Reply:
I'm not sure I really understand the slice configuration. It seems ambiguous in many cases and so there are scenarios where it doesn't work the same way. For example, what if the user puts the slice 0 configuration in KAFKA_BROKER_CONFIG then they add the slice 1 configuration in SLICED_KAFKA_BROKER_CONFIG. If that is not valid, it should be validated and prevented but there is nothing preventing the user from doing that now.

Now, I would expected passing slice=0 to fall back to the KAFKA_BROKER_CONFIG but it does not. Instead it goes back to BROKER_CONFIG. This is pretty unexpected to me.

Reply:
To add some context, an example of how the slicing convention would work is here https://github.com/getsentry/ops/blob/master/k8s/services/snuba/templates/prod_settings.py#L413-L537.

To the point of slice=0 behaving same as slice=None, it makes things implicit. I prefer explicit over implicit. So the slicing configuration for different objects which would get sliced would have their own configuration. What slice 0 has in the sliced configs is same as what the original config has. You can see that in the settings file I linked above.

It also keeps thing simpler to reason about. If slicing is enabled on an object(broker config, topic, clusters), look at the sliced version of the config. If not, look at the regular config.

Reply:
If that's how settings is supposed to be defined, this is ok. But this was not clear to me. If sliced settings cannot be put into the KAFKA_BROKER_CONFIG then this needs to be clear from:
- the comments in the settings file
- properly validated in https://github.com/getsentry/snuba/blob/master/snuba/settings/validation.py so it will be impossible to provide such settings

Otherwise, I would assume it might work in the other way where slice 0 can go in KAFKA_BROKER_CONFIG and other slices can go in SLICED_KAFKA_BROKER_CONFIG or that passing slice=0 is fine if there's only 1 slice / unsliced.

Reply:
I see where the confusion is occurring. I can make the comments clearer (although it does say, within the slicing section in settings/init, that each of those configs are for sliced topics/storages/etc only). I'll add some validation as well so that it is impossible to incorrectly configure.

Reply:
That behaviour seems good to me.

I think maybe the settings for KAFKA_BROKER_CONFIG just needs a comment saying that broker config for sliced storages can not go in there.

I think we should error in 2 cases:
- if we see the topic for a sliced storage in in KAFKA_BROKER_CONFIG
- an unsliced topic present in SLICED_KAFKA_BROKER_CONFIG.

Since this is sort of tangential to this particular PR though, feel free to tackle it separately.

Reply:
That sounds good. I'll make this in a separate PR.